Add support for migrations (#83)
sindresorhus/conf: readme.md

> # conf [![Build Status](https://travis-ci.org/sindresorhus/conf.svg?branch=master)](https://travis-ci.org/sindresorhus/conf)

Reviewer: Can you keep this line?

Author: Yes, of course. I have no idea why this got deleted 😕

**Note:** The `default` value will be overwritten by the `defaults` option if set.

### migrations

Type: `Object`

You can use migrations to perform operations to the store whenever a version is switched.

The `migrations` object should be consisted of a key-value pair of `version`: `handler`.

**Note**: The [`projectVersion`](#projectversion) option should be specified in order for the migrations to be run.

Example:

```js
const store = new Conf({
    migrations: {
       '0.0.1': store => {
            store.set('debug phase', true);
       },
       '1.0.0': store => {
            store.delete('debug phase');
            store.set('phase', '1.0');
       },
       '1.0.2': store => {
            store.set('phase', '>1.0');
       }
    }
});
```

> 

Reviewer 2: Same feedback as your electron-store PR.

Author: Done 😄

**Note:** The `default` value will be overwritten by the `defaults` option if set.

### migrations

Type: `Object`

You can use migrations to perform operations to the store whenever a version is switched.

The `migrations` object should be consisted of a key-value pair of `version`: `handler`.

**Note**: The `projectVersion` option should be specified in order for the migrations to be run.

> But you only need to specify the version if you don't have a package.json file in your project.

Reviewer: Perhaps change this line to emphasize that we assume the value of the `version` key. The `version` field might not be set in the package.json.

> You only need to specify the `projectVersion` if there's no version defined in the project's package.json file or there's no package.json at all.

Reviewer: Also, I think it's worth documenting the `projectVersion` option, after the `projectName`  one.

Author: Agree! Nice catch.

> But you only need to specify the `projectVersion` if you don't have a package.json file in your project.

Reviewer: Perhaps remove this line which is similar to what is described in the option itself and link `projectVersion` to the option itself in the sentence above (The **`projectVersion`** option...)?

Author: Done, seems better now?

Reviewer: Yep, but I think you have missed documenting that handlers must be idempotent and briefly explain why (they will run more than once if migrations fail).

Author: > > Is it possible to use, for example, store.set('**internal**.version', '1.2.3') ?
> 
> Do you mean for users? Then yes, but it should not be allowed. We should document that the `__internal__` key is reserved and throw if it's used from the outside. I doubt anyone would ever encounter it though.

I mean for internal usage. If the user has set `accessPropertiesByDotNotation` to false, the internal migration code would fail, right?

Author: Do you want to wait for the discussion in https://github.com/sindresorhus/conf/pull/83#discussion_r306199633 to proceed before I apply the change you suggested?

Reviewer: Lets wait for @sindresorhus feedback. Still, there are situations in which the function might run twice, even if we make the effort of prevent most of them. So the comment still makes sense to be added here.

Type: `string`<br>
Default: The `name` field in the package.json closest to where `conf` is imported.

> You only need to specify this if you don't have a package.json file in your project or if there isn't a name defined within it.

Reviewer 2: ```suggestion
You only need to specify this if you don't have a package.json file in your project or if it doesn't have a `"name"` field.
```

Author: Done, also replaced it in other 3 places.

You only need to specify this if you don't have a package.json file in your project.

#### projectVersion

Type: `string`<br>
Default: The `version` field in the package.json closest to where `conf` is imported.

> You only need to specify this if you don't have a package.json file in your project.

Reviewer: You only need to specify this if you don't have a package.json file in your project or if there isn't a version defined within it.

Author: Done. I did the same for the name!
